CLAUDE.md

```diff
@@ -0,0 +1,529 @@
+
+For operations that need type specialization, templates in `csrc/*.jinja` can generate C++ code:
+
+```jinja
+// Input template
+using DTypeIn = {{ dtype_in }};
+using DTypeOut = {{ dtype_out }};
+constexpr int PARAM = {{ param_value }};
+
+// After render
+using DTypeIn = float16;
+using DTypeOut = float16;
+constexpr int PARAM = 128;
+```
+
+This allows the same CUDA template code to be compiled with different concrete types. However, if your operation doesn't need this, you can skip Jinja and write the `.cu` files directly.
+
+## Directory Structure
+
+```
+flashinfer/
+├── include/flashinfer/           # Header-only CUDA kernel templates
+│   ├── attention/                # Attention kernels
+│   ├── gemm/                     # GEMM kernels
+│   ├── comm/                     # Communication kernels
+│   ├── mma.cuh                   # Matrix multiply utilities
+│   ├── utils.cuh                 # Common utilities
+│   └── [...]
+│
+├── csrc/                          # Framework bindings (via TVM-FFI)
+│   ├── *.cu                       # Kernel launcher implementations
+│   ├── *_jit_binding.cu           # TVM-FFI exports
+│   ├── *_customize_config.jinja   # Type config templates (optional)
+│   └── [...]
+│
+├── flashinfer/                    # Python package
+│   ├── jit/
+│   │   ├── core.py                # JitSpec, compilation infrastructure
+│   │   ├── cpp_ext.py             # Ninja build generation
+│   │   ├── env.py                 # Workspace paths
+│   │   ├── attention/             # Attention module generators
+│   │   ├── gemm/                  # GEMM module generators
+│   │   ├── fused_moe/             # MOE module generators
+│   │   └── [...]
+│   ├── *.py                       # High-level Python APIs
+│   ├── aot.py                     # AOT compilation for pre-built packages
+│   └── [...]
+│
+└── build_backend.py               # PEP 517 build backend
+```
+
+### Critical Rule: Framework Separation
+
+**Torch headers MUST NOT be included in `include/` directory files.**
+
+- `include/`: Framework-agnostic CUDA kernels (accept raw pointers)
+- `csrc/`: Framework bindings via TVM-FFI (currently PyTorch, but can support other frameworks)
+
+## Adding a New Operation
+
+→ **For complete step-by-step tutorial, see [`.claude/skills/add-cuda-kernel/skill.md`](.claude/skills/add-cuda-kernel/skill.md)**
+
+**Quick overview of the process:**
+1. Write kernel in `include/flashinfer/new_op.cuh` (framework-agnostic, raw pointers)
+2. Write launcher in `csrc/new_op.cu` (PyTorch tensor handling)
+3. Create TVM-FFI bindings in `csrc/new_op_jit_binding.cu`
+4. (Optional) Create Jinja template for type specialization
+5. Write JIT module generator in `flashinfer/jit/new_op.py`
+6. Write Python API in `flashinfer/new_op.py` with `@functools.cache`
+7. Write tests in `tests/`
+8. Register in `flashinfer/aot.py` for AOT compilation
+9. Export in `flashinfer/__init__.py`
+
+**Reference implementations:**
```

Comment:
Example implementations.

Reference sounds like it could be used for accuracy testing.